packages/Takeover/src/Takeover.js

```diff
@@ -0,0 +1,116 @@
+import React from "react";
+import ReactDOM from "react-dom";
+import PropTypes from "prop-types";
+import { Transition } from "react-transition-group";
+import FocusTrapLibrary from "focus-trap-react";
+import styled from "styled-components";
+import EscListener from "./components/EscListener";
+import Wrapper from "./components/Wrapper";
+import Header from "./components/Header";
+import LockBodyScroll from "./components/LockBodyScroll";
+import { animationDuration } from "./tokens";
+import extractChildren from "./helpers/extractChildren";
+
+const propTypes = {
+  /** The content for the Takeover. */
+  children: PropTypes.node.isRequired,
+
+  /** Control the visibility of the takeover */
+  isOpen: PropTypes.bool.isRequired,
+
+  /** Callback triggered when the takeover needs to be close */
+  onClose: PropTypes.func,
+
+  /** Callback once the takeover has been opened event */
+  onAfterOpen: PropTypes.func,
+
+  /** Callback once the takeover has been closed event */
+  onAfterClose: PropTypes.func,
+
+  /** Render the takeover inline */
+  isInline: PropTypes.bool,
+
+  /** Takeover will capture and will not propagate event after pressing ESC. Helps in case we have many overlapped components. */
+  shouldStopEscapePropagation: PropTypes.bool,
+};
+
+const defaultProps = {
+  isInline: false,
+  shouldStopEscapePropagation: false,
+  onAfterClose: () => {},
+  onClose: () => {},
+  onAfterOpen: () => {},
+};
+
+const Portal = ({ children, active }) =>
+  active ? ReactDOM.createPortal(children, document.body) : <React.Fragment>{children}</React.Fragment>;
+
+const StyledWrapper = styled(Wrapper)`
+  display: flex;
+  flex-direction: column;
+`;
+
+const StyledHeader = styled(Header)`
+  flex: none;
+`;
+
+const ContentWrapper = styled.div`
+  overflow-y: auto;
+  flex: 1 1 auto;
+`;
+
+const Takeover = ({ isOpen, onClose, isInline, shouldStopEscapePropagation, onAfterClose, onAfterOpen, ...props }) => {
+  const refWrapper = React.useRef(null);
+
+  function handleTransitionEnter(node) {
+    // https://github.com/reactjs/react-transition-group/blob/6dbadb594c7c2a2f15bc47afc6b4374cfd73c7c0/src/CSSTransition.js#L44
+    node.scrollTop;
+  }
+
+  const {
+    "Takeover.FocusTrap": focusTrapExtracted,
+    "Takeover.Header": headerExtracted,
+    "Takeover.Content": contentExtracted,
+    children,
+  } = extractChildren(props.children, ["Takeover.FocusTrap", "Takeover.Header", "Takeover.Content"]);
+
+  const extendedFocusTrapOptions = focusTrapExtracted ? focusTrapExtracted.props : {};
+
+  const focusTrapOptions = {
+    fallbackFocus: () => document.createElement("div"),
+    ...extendedFocusTrapOptions,
+  };
+
+  return (
+    <>
+      {isOpen && <EscListener on={onClose} shouldStopEscapePropagation={shouldStopEscapePropagation} />}
+      {isOpen && <LockBodyScroll />}
+      <Portal active={!isInline}>
+        <Transition
```

Comment:
Can we extract maybe this component `<Transition />` as well so we can reuse it in modal, sidepanel and takeover 🤔?

Reply:
IMHO, modal, takeover and sidepanel is a one component with different props:
Modal: `<Dialog>...</Dialog>`
Takeover: `<Dialog align="stretch">...</Dialog>`
Sidepanel: `<Dialog align="right">...</Dialog>`

So I suggest to not extract `<Transition />`, but at some point create `Dialog` component that combines modal, takeover and sidepanel functionality. (And deprecate Sidepanel and Takeover after that)

I have task about Modal component in backlog. Probably, it would be better to implement common Dialog, than separate Modal.

Reply:
Ok, let's start from implementing Modal as separated component. I will try to extract `<Transition />` later when start working on Modal component.